// class-enum/dist/ClassEnum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const exception_1 = require("./exception");
class ClassEnum {
    constructor(value) {
        this.value = value;
    }
    static values() {
        const enums = this.getEnums();
        return Object.keys(enums).map(function (name) {
            return enums[name];
        });
    }
    static getEnums() {
        const classEnums = {};
        for (const name of Object.getOwnPropertyNames(this)) {
            if (name === 'prototype') {
                continue;
            }
            const descriptor = Object.getOwnPropertyDescriptor(this, name);
            if (!descriptor || !(descriptor.value instanceof this)) {
                continue;
            }
            // @ts-ignore
            classEnums[name] = descriptor.value;
        }
        return classEnums;
    }
    static valueOf(value, defaultEnum = null) {
        const enums = this.getEnums();
        if (!(value in enums)) {
            if (defaultEnum !== null) {
                return defaultEnum;
            }
            throw new exception_1.EnumNotFound(value);
        }
        return enums[value];
    }
    name() {
        return this.value;
    }
    equals(e) {
        return this === e && this.value === e.value;
    }
}
exports.default = ClassEnum;
